ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh

```diff
@@ -0,0 +1,139 @@
+        oc adm upgrade --to-image="${target}" --allow-explicit-upgrade
+        echo "Upgrading cluster to ${target} gets started..."
+    else       
+        oc adm upgrade --to-image="${target}" --allow-explicit-upgrade --force
+        echo "Force upgrading cluster to ${target} gets started..."       
+    fi
+    return 0
+}
+
+# Monitor the upgrade status
+function check_upgrade_status() {
+    local wait_upgrade="${TIMEOUT}" out avail progress
+    while (( wait_upgrade > 0 )); do
+        echo "oc get clusterversion" && oc get clusterversion
+        out="$(oc get clusterversion --no-headers)"
+        avail="$(echo "${out}" | awk '{print $3}')"
+        progress="$(echo "${out}" | awk '{print $4}')"
+        if [[ ${avail} == "True" && ${progress} == "False" && ${out} == *"Cluster version is ${target_version}" ]]; then
+            echo -e "Upgrade succ\n\n"
+            return 0
+        else
+            sleep 5m
+            (( wait_upgrade -= 5 ))
+        fi        
+    done
+    if (( wait_upgrade <= 0 )); then
+        echo >&2 "Upgrade timeout, exiting" && return 1
+    fi
+}
+
+# Check version, state in history
+function check_history() {
+    local cv version state
+    cv=$(oc get clusterversion/version -o json)
+    version=$(echo "${cv}" | jq -r '.status.history[0].version')
+    state=$(echo "${cv}" | jq -r '.status.history[0].state')
+    if [[ ${version} == "${target_version}" && ${state} == "Completed" ]]; then
+        echo "History check PASSED, cluster is now upgraded to ${target_version}" && return 0
+    else
+        echo >&2 "History check FAILED, cluster upgrade to ${target_version} failed, current version is ${version}, exiting" && return 1
+    fi
+}
+
+# Setup proxy if it's present in the shared dir
+if test -f "${SHARED_DIR}/proxy-conf.sh"
+then
+    # shellcheck disable=SC1091
+    source "${SHARED_DIR}/proxy-conf.sh"
+fi
+
+# Note: RELEASE_IMAGE_xxx is by-digest pull spec
+# oc registry login creates registry credentials and then the oc adm release info command works 
+oc registry login
+
+echo "RELEASE_IMAGE_INITIAL is ${RELEASE_IMAGE_INITIAL}"
+echo "RELEASE_IMAGE_LATEST is ${RELEASE_IMAGE_LATEST}"
+
+echo -e "Current cluster version: oc get clusterversion\n"
+oc get clusterversion
+
+echo -e "RELEASE_IMAGE_INITIAL release info:\n"
+oc adm release info "${RELEASE_IMAGE_INITIAL}"
+
+echo -e "RELEASE_IMAGE_LATEST release info:\n"
+oc adm release info "${RELEASE_IMAGE_LATEST}"
+
+target="${RELEASE_IMAGE_LATEST}"
+
+source_minor_version="$(oc get clusterversion --no-headers | awk '{print $2}' | cut -f2 -d.)"
+echo -e "Source release minor version is: ${source_minor_version}"
+
+target_version="$(oc adm release info "${target}" --output=json | jq -r '.metadata.version')"
+target_minor_version="$(echo "${target_version}" | cut -f2 -d.)"
+echo -e "Target release version is: ${target_version}\nTarget minor version is: ${target_minor_version}"
+
+if [[ "${FORCE_UPDATE}" == "false" ]]; then
+    admin_ack 
```

Comment:
Thanks for catching it. We set set -o errexit at the beginning, so by default when admin_ack gets a non-zero return value, the script will exit unless when explicitly check the value.

Reply:
ack

Reply:
Happen to think about we need one more check that if the target payload signed or not before do upgrade without --force.

Reply:
I didn't add the check because CVO would prompt the error. But yeah, if we do this check, the script could exit earlier when updating to an unsigned build w/o force. Ack. Will do. Thanks.